### ISSUE TITLE
Feature request: Ability to correct global offenses

### ISSUE THREAD

**Issue by Zopolis4** [rubocop/rubocop#13045]
## Is your feature request related to a problem? Please describe.

```ruby
      # Adds an offense that has no particular location.
      # No correction can be applied to global offenses
      def add_global_offense(message = nil, severity: nil)
```

In my extension, I have a cop which returns global offenses, but is able to autocorrect them. I can achieve functioning autocorrect by using `autocorrect_requested?`, but this does not report back to the user as correcting the offense.

## Describe the solution you'd like

Either the ability to pass a block to `add_global_offense` in the same way that `add_offense` does, or just a way to manually mark an autocorrect as having being completed.

## Describe alternatives you've considered

I have considered using `add_offense` with a dummy comment value, (i.e. a slight modification of how `Lint/ScriptPermission` does it, but given that this is a global offense returning a fake comment seemed less than ideal.

## Additional context

```ruby
module RuboCop
  module Cop
    module Crystal
      # Non-executable ruby files commonly have an extension of `.rb`.
      # The equivalent crystal files have an extension of `.cr`.
      #
      # @example
      #   # bad
      #   example.rb
      #
      #   # good
      #   example.cr
      #
      class FileExtension < Base
        extend AutoCorrector
        MSG = 'Crystal files have `.cr` extensions, while Ruby files have `.rb` extensions.'

        def on_new_investigation
          return unless File.extname(processed_source.file_path) == '.rb'

          add_global_offense
          File.rename(processed_source.file_path, File.join(File.dirname(processed_source.file_path), "#{File.basename(processed_source.file_path, '.rb')}.cr")) if autocorrect_requested?
        end
      end
    end
  end
end
```

```
$ ls
example.rb
$ rubocop --require rubocop-crystal -A --only Crystal/FileExtension
Inspecting 1 file
C

Offenses:

example.rb:1:1: C: Crystal/FileExtension: Crystal files have .cr extensions, while Ruby files have .rb extensions.

1 file inspected, 1 offense detected
$ ls
example.cr
```

**Comment by dvandersluis** [comment]
Technically autocorrect relates to rewriting the AST, which isn't happening in your example. In the cases I looked at of using `add_global_offense`, including yours, there aren't really any instances where there's an AST correction to be made.

I expect renaming a file in the middle of a run will have side effects (eg. the files being acted on are collected before RuboCop starts).

This is probably why for example [`RSpec/SpecFilePathSuffix`](https://github.com/rubocop/rubocop-rspec/blob/master/lib/rubocop/cop/rspec/spec_file_path_suffix.rb) (which is a parallel check I think) doesn't have autocorrect.

**Comment by Zopolis4** [comment]
I haven't noticed any problems using my current method of renaming the file as part of the cop, although I suppose my extension only has a few cops and I haven't run it on many files. `Lint/EmptyFile` only supports non-autocorrectable global offenses, although presumably you could autocorrect the offense by just deleting the file? Then again, that's also not an AST correction.

**Comment by Zopolis4** [comment]
This issue is popping up again after https://github.com/rubocop/rubocop/pull/15430, as that now breaks `Crystal/FileExtension`. 

I can avoid an error by using the same `File.exist?(processed_source.file_path)` workaround as https://github.com/rubocop/rubocop/pull/15222, but the files will get renamed by `Crystal/FileExtension` but then also written back to disk as `.rb` files by other cops that have their own corrections.

This is fixed if I manually run the cop twice, ~~or if I add a dummy `add_offense`, but if `add_global_offense` supported autocorrections then I'd just be able to use the same fix as `Lint/ScriptPermission` and not have to change anything else. (Presumably `Lint/ScriptPermission` has the same issue, but because it uses an autocorrectable regular offense, it is caught on a second run)~~ I'm not able to reproduce this behavior using an `add_offense` anymore.
